### PR TITLE
Add the tag list page to the release page

### LINKED ISSUE #12081: Cannot delete tags
- Gitea version (or commit ref): 1.12
- Git version: ?
- Operating system: Linux
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [ ] Not relevant
- Log gist:

## Description
Once a tag has been pushed, it cannot be removed anymore. Gitea doesn't offer that from the UI (or I didn't find where) – and deleting the tag locally plus using `git push --tags` doesn't remove it either.

Left some of the check-marks open as I'm not sure what Codeberg is running (@lunny probably knows, else @ashimokawa can tell if it's needed).

## Screenshots
cannot screenshot what's not there (oops), sorry.

### PR DESCRIPTION
**Pull request by a1012112796**
* Add the tags list view
* Add the delete tag way on ui
* Not delete tag and clear message when delete a release
* Not show single tag in release list
example view:

tmp1

tmp2

### CODE REVIEW

**Comment by UnlimitedCookies** [comment]
Looks awesome! How about putting a trashcan symbol instead of text in brackets behind the Release version?

**Review by zeripath — APPROVED** [review]

**Comment by mrsdizzie** [comment]
This doesn't delete a tag from the git repo it just removes reference from the releases table. This means the tag still exists, you can still view it in tags list on main repo view, and still create a release from it. If releases for repo get re-synced it will show up on the releases page again too. Tags and releases are different you can't use the same code for removing a release -- release only exists in the DB but tag exists within git itself. So this doesn't fixed the linked issue and will probably mislead people into thinking a tag has been deleted when it hasn't. So I think this would have to actually delete the tag in git itself to count.

If it was changed to do actually delete the tag, then I would say:

Agree it should use a trash can instead of text -- Maybe another icon after  TAR.GZ like this:

Screen Shot 2020-07-01 at 9 52 51 AM

I know it was probably to reuse the existing translation, but I think the warning should not use the phrase 'deleting a release' since we make a difference between release and tag and this is about deleting a single tag.

**Comment by a1012112796** [comment]
> This doesn't delete a tag from the git repo it just removes reference from the releases table. This means the tag still exists, you can still view it in tags list on main repo view, and still create a release from it. If releases for repo get re-synced it will show up on the releases page again too. Tags and releases are different you can't use the same code for removing a release -- release only exists in the DB but tag exists within git itself. So this doesn't fixed the linked issue and will probably mislead people into thinking a tag has been deleted when it hasn't. So I think this would have to actually delete the tag in git itself to count.
> 
> If it was changed to do actually delete the tag, then I would say:
> 
> Agree it should use a trash can instead of text -- Maybe another icon after TAR.GZ like this:
> 
> Screen Shot 2020-07-01 at 9 52 51 AM
> 
> I know it was probably to reuse the existing translation, but I think the warning should not use the phrase 'deleting a release' since we make a difference between release and tag and this is about deleting a single tag.

No, I will delete tag in repo also. ref code:
https://github.com/go-gitea/gitea/blob/858c35b73184aa94242076d417ad09a74c55aebf/services/release/release.go#L129-L137

https://github.com/go-gitea/gitea/blob/858c35b73184aa94242076d417ad09a74c55aebf/routers/repo/release.go#L349

for ui, I wll change it.

**Comment by a1012112796** [comment]
> Agree it should use a trash can instead of text -- Maybe another icon after TAR.GZ like this:
> 
> Screen Shot 2020-07-01 at 9 52 51 AM

I don't like to show it on here, which is more like delete attachments..

**Comment by mrsdizzie** [comment]
ah I see I missed that 'true' but the error I see is like this:
```
DeleteReleaseByID: LoadAttributes: user does not exist [uid: 0, name: , keyid: 0]
```
When a repo is migrated it makes 'releases' from existing tags but sets the PublisherID to 0. Then when you try to delete one of those tags it shows the error above.

https://github.com/go-gitea/gitea/blob/858c35b73184aa94242076d417ad09a74c55aebf/services/release/release.go#L153-L155

https://github.com/go-gitea/gitea/blob/858c35b73184aa94242076d417ad09a74c55aebf/models/release.go#L45-L60

Current code is only for releases that are created and assume all release have publisher information, but tags on release page do not.

**Comment by a1012112796** [comment]
> ah I see I missed that 'true' but the error I see is like this:
> 
> ```
> DeleteReleaseByID: LoadAttributes: user does not exist [uid: 0, name: , keyid: 0]
> ```
> 
> When a repo is migrated it makes 'releases' from existing tags but sets the PublisherID to 0. I believe this is true for all tags in the release table. Then when you try to delete one of those tags it shows the error above.
> 
> https://github.com/go-gitea/gitea/blob/858c35b73184aa94242076d417ad09a74c55aebf/services/release/release.go#L153-L155
> 
> https://github.com/go-gitea/gitea/blob/858c35b73184aa94242076d417ad09a74c55aebf/models/release.go#L45-L60
> 
> Current code is only for releases that are created and assume all release have publisher information, but tags on release page do not.

I see, it will also become a bug when delete release ...

**Comment by mrsdizzie** [comment]
Maybe DeleteReleaseByID can check if release.IsTag and not loadAttributes then?

**Comment by a1012112796** [comment]
> Maybe DeleteReleaseByID can check if release.IsTag and not loadAttributes then?

I found an question is that the ``pushUpdateAddTags`` not us user who push tags as release's Publisher but use the author of the commit which is be taged. I wonder whether it's right?
https://github.com/go-gitea/gitea/blob/858c35b73184aa94242076d417ad09a74c55aebf/modules/repository/update.go#L56-L110

**Comment by UnlimitedCookies** [comment]
On the UI end:
> > Agree it should use a trash can instead of text -- Maybe another icon after TAR.GZ like this:
> > Screen Shot 2020-07-01 at 9 52 51 AM
> 
> I don't like to show it on here, which is more like delete attachments..

I agree the trashcan symbol needs to visually stand out from the release attachments.
Why not place it behind the version number (where the text previously was)?
And to make it stand out more maybe color it (red?) and change the size.

**Comment by lunny** [comment]
So if somebody wants only to remove release but not tag, how should he do?

**Comment by a1012112796** [comment]
> So if somebody wants only to remove release but not tag, how should he do?

Yes, the current logic isn't reasonable, and we should provide the option to choose whether to delete the tag when you delete the release, but that's not the aim of this PR.

**Comment by lunny** [comment]
How about a new tab named tags on the home page? And user could manage tags there.

**Comment by a1012112796** [comment]
> How about a new tab named tags on the home page? And user could manage tags there.

Maybe can use same style with github:
image

**Comment by lunny** [comment]
I would like to replace `25MiB` as `tags`.
图片

**Comment by a1012112796** [comment]
This PR has been changed. Please review it again. Thanks.

**Comment by CirnoT** [comment]
> I would like to replace `25MiB` as `tags`.
> 图片

That information is already available a little higher, on Releases tab. On the other hand, the size is not visible anywhere else.

**Comment by UnlimitedCookies** [comment]
> > I would like to replace `25MiB` as `tags`.
> > 图片
> 
> That information is already available a little higher, on Releases tab. On the other hand, the size is not visible anywhere else.

Why not add the number of tabs after the number of commits and when you click on tags it will show a list of tags similar to the commit list? 

This would mean that the list would look like this: `1477 Commits                  45 Tags                      25 Branches                     25 MiB`

File: options/locale/locale_en-US.ini
Comment (CirnoT) [inline options/locale/locale_en-US.ini]:
```suggestion
release.deletion_desc = Deleting a release only removes it from Gitea. Git tag, repository contents and history remain unchanged. Continue?
```

File: options/locale/locale_en-US.ini
Comment (CirnoT) [inline options/locale/locale_en-US.ini]:
```suggestion
release.deletion_tag_desc = Will delete this tag from repository. Repository contents and history remain unchanged. Continue?
```

**Comment by silverwind** [comment]
Are the OP screenshots up to date? I can spot various minor layout issues. As for the "tabs" on the release page, maybe we can also do pill-style tabs like GitHub, not sure if Fomantic has such elements natively thought.

File: routers/repo/release.go
Comment (zeripath) [inline routers/repo/release.go]:
```suggestion
		IncludeTags:   isTagList,
```

File: routers/repo/release.go
Comment (zeripath) [inline routers/repo/release.go]:
```suggestion
```

File: routers/repo/release.go
Comment (zeripath) [inline routers/repo/release.go]:
```suggestion
	if tagName := ctx.Query("tag"); len(tagName) > 0 {
```

File: options/locale/locale_en-US.ini
Comment (zeripath) [inline options/locale/locale_en-US.ini]: This doesn't appear to be used?

Comment (a1012112796) [reply options/locale/locale_en-US.ini]: Sorry, I forgot to remove it.

**Comment by silverwind** [comment]
New screenshots look much better. I'd suggest removing the horizontal divider on the first screenshots. Will test this later.

**Comment by CirnoT** [comment]
The "New Release" and "Release detail" looks weird.

**Comment by lunny** [comment]
Why I prefer to move tags list under code but not `releases`. Because sometimes, releases will only include binaries but not codes. 
I don't think we should follow github's design here.

**Comment by a1012112796** [comment]
> Why I prefer to move tags list under code but not `releases`. Because sometimes, releases will only include binaries but not codes.
> I don't think we should follow github's design here.

@lunny Do you mean this? Thanks. 
631cf722554ea1b25f18636f0aeaba0c569f4043

**Comment by lunny** [comment]
@a1012112796 yes.

File: templates/repo/release/list.tmpl
Comment (CirnoT) [inline templates/repo/release/list.tmpl:22]: The table should be hidden if no tags are present, similar to how releases one is hidden

chrome_2020-09-15_14-13-29

**Review by lafriks — CHANGES_REQUESTED** [review]
Tag deletion should be allowed only if user has write rights to code unit

File: templates/repo/release/list.tmpl
Comment (silverwind) [inline templates/repo/release/list.tmpl]: Can you put a bit of horizontal margin around these links? You can add `class="mx-2"` for example to have a horizontal margin of .5rem around them.

Comment (silverwind) [reply templates/repo/release/list.tmpl]: Also, SVG size attribute is no longer required since https://github.com/go-gitea/gitea/pull/12814, so you can change to format `{{svg "octicon-tag"}}`.

Also instead of `nbsp;`, add margin directly on the svg via `{{svg "octicon-tag" 16 "mr-2"}}` (note size is required if you specifify class attributes on the svg helper).

File: options/locale/locale_en-US.ini
Comment (lafriks) [inline options/locale/locale_en-US.ini]:
```suggestion
release.detail = Release details
```

**Comment by silverwind** [comment]
Please pull in https://github.com/silverwind/gitea/commit/f90cabf214f9c4017d845bc3b3d0331c4a76ef13.

File: templates/repo/release/list.tmpl
Comment (silverwind) [reply templates/repo/release/list.tmpl]: Did all suggestions in https://github.com/go-gitea/gitea/pull/12096#issuecomment-695789131.

**Comment by silverwind** [comment]
Also pull in https://github.com/silverwind/gitea/commit/333f6bd9531f52411a9e9798ea9b8ade36079e2c for monospace hashes:

image

**Comment by silverwind** [comment]
I don't think we should list releases on the tags tab. It's just confusing to have them there and GitHub also does not do that.

**Review by silverwind — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
:rocket:

**Comment by lafriks** [comment]
Oh, this is already moved to 1.14.0

**Comment by Fowlthane** [comment]
Hello everyone,

This new feature is fantastic and I'm looking forward to it.

would it be possible to have this new feature in 1.13?

**Thank you!**

**Review by lunny — DISMISSED** [review]
Users could view the tag list page should have UnitCode read permission.

**Comment by techknowlogick** [comment]
@a1012112796 please resolve conflicts 
@lunny please review again

**Comment by lunny** [comment]
And once somebody have no `unitcode` read permission. The `tags` tab should be hidden.

File: web_src/less/_base.less
Comment (a1012112796) [inline web_src/less/_base.less]: @silverwind I found ``@monospaced-fonts`` is not defined after merge master, Please have a look. Thanks
log:
```SHELL
ERROR in ./web_src/less/index.less
Module build failed (from ./node_modules/mini-css-extract-plugin/dist/loader.js):
ModuleBuildError: Module build failed (from ./node_modules/less-loader/dist/cjs.js):


  .commit-sha {
    font-family: @monospaced-fonts, monospace;
               ^
Variable @monospaced-fonts is undefined
      Error in /gitea_work/gitea/web_src/less/_base.less (line 544, column 17)
    at /gitea_work/gitea/node_modules/webpack/lib/NormalModule.js:316:20
    at /gitea_work/gitea/node_modules/loader-runner/lib/LoaderRunner.js:367:11
    at /gitea_work/gitea/node_modules/loader-runner/lib/LoaderRunner.js:233:18
    at context.callback (/gitea_work/gitea/node_modules/loader-runner/lib/LoaderRunner.js:111:13)
    at Object.lessLoader (/gitea_work/gitea/node_modules/less-loader/dist/index.js:57:5)
 @ multi ./web_src/js/jquery.js ./web_src/fomantic/build/semantic.js ./web_src/js/index.js ./web_src/fomantic/build/semantic.css ./web_src/less/index.less index[4]
```

Comment (silverwind) [reply web_src/less/_base.less:544]: You need to use `font-family: var(--fonts-monospace);` now.

File: web_src/less/_base.less
Comment (silverwind) [inline web_src/less/_base.less]:
```suggestion
    font-family: var(--fonts-monospace);
```

File: web_src/less/_base.less
Comment (silverwind) [reply web_src/less/_base.less]: See suggestion below.

**Comment by silverwind** [comment]
Good to go? (after rebase)

**Comment by lafriks** [comment]
tests are failing:
```
release_test.go:59: 
        	Error Trace:	release_test.go:59
        	            				release_test.go:104
        	Error:      	Not equal: 
        	            	expected: 3
        	            	actual  : 2
        	Test:       	TestCreateReleaseDraft
```

**Comment by a1012112796** [comment]
> tests are failing:
> 
> ```
> release_test.go:59: 
>         	Error Trace:	release_test.go:59
>         	            				release_test.go:104
>         	Error:      	Not equal: 
>         	            	expected: 3
>         	            	actual  : 2
>         	Test:       	TestCreateReleaseDraft
> ```

It's not a problem, because the new item added by #13358 is a tag, which will not be shown in the release list, so the actual number is right :)

https://github.com/go-gitea/gitea/blob/543697e61e4881256ab56a0fd26eed9765eac1de/models/fixtures/release.yml#L32-L45

**Comment by a1012112796** [comment]
The ui is broken after merge master. @silverwind  Would you please have a look , Thanks.
image

**Comment by silverwind** [comment]
Please pull in https://github.com/silverwind/gitea/commit/74a04ac048d6adb1cd4431bdf15ff6702669145e which fixes those and does a few more enhancements:

Screen Shot 2020-11-01 at 17 54 10
Screen Shot 2020-11-01 at 17 53 55
Screen Shot 2020-11-01 at 17 58 01
Screen Shot 2020-11-01 at 17 58 13

**Comment by silverwind** [comment]
Also pull in https://github.com/silverwind/gitea/commit/71aa21e1715f76acac55144e8bd4072f0739369a for some more tweaks on the title:

Screen Shot 2020-11-01 at 18 04 21

**Comment by lafriks** [comment]
@lunny need your review

File: templates/repo/release/list.tmpl
Comment (silverwind) [inline templates/repo/release/list.tmpl]:
```suggestion
										<a class="mr-3 mono" href="{{$.RepoLink}}/src/commit/{{.Sha1}}" rel="nofollow">{{svg "octicon-git-commit" 16 "mr-2"}}{{ShortSha .Sha1}}</a>
```

File: web_src/less/_base.less
Comment (silverwind) [inline web_src/less/_base.less]:
```suggestion
```

**Comment by lafriks** [comment]
🚀

**Pull request merged**
